### PR TITLE
New features for repeating rows and columns in the print setup

### PR DESCRIPTION
**Pull request by dorssar**
Added two private attributes to Worksheet - repeatingRows and repeatingCols.

Some examples how it can be used:

ws.repeatRows(1, 5) -- will keep first top 5 rows on each page when printing.
ws.repeatRows(2) -- will keep the second row (marked 2) on each page when printing.

ws.repeatCols("A", "C") -- will repeat columns A to C (A, B, and C) on each page when printing.
ws.repeatCols("A") -- will repeat only column A on each page for the print setup.

The private attributes repeatingRows and repeatingCols are ArrayLists, so functions repeatCols and repeatRows are handling their conversion to a proper ArrayList format.

The user can call repeatRows countless times, but only the last call will count. (same applies to repeatCols)

Workbook then takes care of defining these features per each sheet.

### CODE REVIEW

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java [hunk 1]
```diff
@@ -191,7 +192,41 @@ private void writeWorkbookFile() throws IOException {
...
+                        w.append("<definedNames>");
+                        w.append("<definedName function=\"false\" " + 
+                                 "hidden=\"false\" " +
+                                 "localSheetId=\"0\" " + 
+                                 "name=\"_xlnm.Print_Titles\" " + 
+                                 "vbProcedure=\"false\">");
+                        w.append(defineName);
+                        w.append("</definedName>");
+                        w.append("</definedNames>");
+                    } catch (IOException e) {
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java:223]: This method declares `throws IOException`, so don't catch it here. Just remove the `try` -  `catch`

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java [hunk 2]
```diff
@@ -83,11 +78,6 @@
      * Map of columns and their widths
      */
     private final Map<Integer, Double> colWidths = new HashMap<>();
-
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java:86]: Was there a problem with having custom row heights and repeating rows? Nevertheless don't just remove existing functionality.

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java [hunk 3]
```diff
@@ -918,6 +915,30 @@ public void pageOrientation(String orientation) {
         this.pageOrientation = orientation;
     }
 
+    public void repeatRows(int startRow, int endRow) {
+        this.repeatingRows = new ArrayList<Integer>();
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java:919]: Use a specific class to store a range. Like 
```
private static class RepeatRange { 
  final int start;
  final int end;
}
```

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java [hunk 4]
```diff
@@ -918,6 +915,30 @@ public void pageOrientation(String orientation) {
...
+    }
+
+    public void repeatRows(int row) {
+        this.repeatingRows = new ArrayList<Integer>();
+        this.repeatingRows.add(row);
+        this.repeatingRows.add(row);
+    }
+
+    public void repeatCols(String startCol, String endCol) {
+        this.repeatingCols = new ArrayList<String>();
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java:931]: fastexcel's API uses integer column indexes everywhere else. 
Use `Range.colToString(int)` to convert column index to column name where writing to xml.

**Review by rzymek — CHANGES_REQUESTED** [review]

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java (hunk 2 above)
Comment (dorssar) [reply fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java:86]: Yes, sorry, this was an accident

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java (hunk 1 above)
Comment (dorssar) [reply fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java:223]: That will not work, it still requires handling an exception inside the forEach streams (https://stackoverflow.com/questions/29838774/unhandled-ioexception-inside-java-list-stream-foreach-lambda-expression)
Maybe I should avoid forEach then

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java (hunk 4 above)
Comment (dorssar) [reply fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java:964]: Makes sense, will fix that.

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java (hunk 1 above)
Comment (rzymek) [reply fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java:223]: Yes, use standard `for` loop instead of `forEach()`

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java (hunk 3 above)
Comment (dorssar) [reply fastexcel-writer/src/main/java/org/dhatim/fastexcel/Worksheet.java]: made it a public class, since it also needs to be used outside of the Worksheet class (in the Workbook class)

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java [hunk 5]
```diff
@@ -191,7 +192,36 @@ private void writeWorkbookFile() throws IOException {
...
+                RepeatRange repeatingCols = ws.getRepeatingCols();
+                RepeatRange repeatingRows = ws.getRepeatingRows();
+
+                if (repeatingCols != null) {
+                    defineName += ws.getName() + "!" + repeatingCols.colRangeToString();
+                }
+                if (repeatingRows != null) {
+                    if (repeatingCols != null) defineName += ","; //add description separator
+                    defineName += ws.getName() + "!" + repeatingRows.rowRangeToString();
+                }
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java:211]: How about:
```
String defainedName = Stream.of(ws.getRepeatingCols(),ws.getRepeatingRows())
  .filter(nonNull)
  .map(r=>ws.getName()+"!"...)
  .collect(joining(",")
```

**Review by rzymek — CHANGES_REQUESTED** [review]

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java (hunk 5 above)
Comment (dorssar) [reply fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java]: improved!

File: README.md [hunk 6]
```diff
@@ -121,6 +121,14 @@ To freeze the first column on the left and the top three rows:
 ```java
 ws.freezePane(1, 3)
 ```
+Define repeating rows or columns for the print setup.
+```java
+ws.repeatRows(1, 5) // keep first top 5 rows on each page when printing.
+ws.repeatRows(2) // keep the second row on each page when printing.
+
+ws.repeatCols("A", "C") // repeat columns A to C (A, B, and C) on each page when printing.
```
Comment (rzymek) [inline README.md:129]: `repeatCols` uses numered indexes now

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Color.java [hunk 7]
```diff
@@ -28,7 +28,7 @@
     public static final String TAN = "D2B48C";
     public static final String PEARL = "EAE0C8";
     public static final String BEIGE = "F5F5DC";
-    public static final String APRICOT = "FBCEB1";
+    public static final String APRICOT = "F6D2B7";
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/Color.java:31]: Why the change in colors?

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/RepeatRangeType.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ */
+package org.dhatim.fastexcel;
+
+/**
+ * Predefined header/footer positions.
+ */
+enum RepeatRangeType {
+    ROW, COLUMN;
+
+    public String getType() {
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/RepeatRangeType.java:24]: Doesn't seem to be used anywhere

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/RepeatRange.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        this.from = from;
+        this.to = to;
+        this.type = type;
+    }
+
+    /**
+     * Column indexes need to be transformed to the letter form.
+     */
+    public String colRangeToString() {
+        return "$" + Range.colToString(from) + ":$" + Range.colToString(to);
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/RepeatRange.java:36]: My bad, I missed that the "definedName" for cols and rows is not identical. 
Instead of differentiating by enum type, you could use different classes:
```
class RepeatRowRange {
final int from; final int to;
String toString() { return "$" + (1 + from) + ":$" + (1 + to); }
}
class RepeatColRange {
final int from; final int to;
String  toString() { return "$" + Range.colToString(from) + ":$" + Range.colToString(to);}
}
```

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/RepeatRange.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+     *  (sheet row indexes start from 1 and not from 0)
+     */
+    public String rowRangeToString() {
+        return "$" + (1 + from) + ":$" + (1 + to);
+    }
+
+    public String toString() {
+        if (type == RepeatRangeType.ROW) return rowRangeToString();
+        else if (type == RepeatRangeType.COLUMN) return colRangeToString();
+        else return null;
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/RepeatRange.java:50]: I don't think toString() should ever return a `null`...

**Review by rzymek — CHANGES_REQUESTED** [review]

File: fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java [hunk 11]
```diff
@@ -184,14 +191,37 @@ private void writeWorkbookFile() throws IOException {
...
+             *  This is defined for each sheet separately 
+             * (if there are any repeating rows or cols in the sheet at all) **/
+    
+            for (Worksheet ws : worksheets) {
+                String defineName = Stream.of(ws.getRepeatingCols(),ws.getRepeatingRows())
+                                .filter(Objects::nonNull)
+                                .map(r -> ws.getName() + "!" + r.toString())
+                                .collect(Collectors.joining(","));
+
+                if (!defineName.equals("")) {
```
Comment (rzymek) [inline fastexcel-writer/src/main/java/org/dhatim/fastexcel/Workbook.java:212]: I'd use `!defineName.isEmpty()`

**Review by rzymek — APPROVED** [review]

**Review by ochedru — APPROVED** [review]
